File: Python/roman_to_integer.py
def romanToInt(rom: str) -> int:
    """Given a roman numeral, convert it to an integer. """

    r = {"I": 1, "V": 5, "X": 10, "L": 50, "C": 100, "D": 500, "M": 1000}
    n = 0
    i = 0
    while i < len(rom):
        rom_key = rom[i]
        num = r[rom_key]
        if i+1 < len(rom):
            next = rom[i+1]
        else:
            next = rom[i]
        if r[next] > r[rom_key]:
            num = r[next] - r[rom_key]
            i += 1
        n = n+num
        i += 1
    return n

File: rust/src/roman_to_integer.rs
use std::{collections::HashMap, char};

pub fn roman_to_int(rom: String) -> i32 {
    /* 
        Given a roman numeral, convert it to an integer.

        Not the prettiest code, but it works. But how to oranize it better?
    */


    
    let romans = HashMap::from([
        ('I', 1),
        ('V', 5),
        ('X', 10),
        ('L', 50),
        ('C', 100),
        ('D', 500),
        ('M', 1000),
    ]);

    let rom_vec: Vec<char> = rom.chars().collect();

    let mut i = 0;
    let mut num = 0;

    while i < rom_vec.len() {
        let value = romans.get(&rom_vec[i]);
        let next_value_option = rom_vec.get(i+1);
        let next_value: Option<&i32>;

        match next_value_option {
            Some(char) => {
                next_value = romans.get(char)
            }
            None => {
                next_value = None;
            }
        }

        match value {
            Some(int) => {
                match next_value {
                    Some(next_int) => {
                        if next_int > int {
                            num = num + (next_int - int);
                            i = i + 2;
                        } else {
                            num = num + int;
                            i = i + 1;
                        }
                    },
                    None => {
                        num = num + int;
                        i = i + 1
                    }
                }
            },
            _ => i = i+1
        }
    }
    num
}